File: 11-container-with-most-water/11-container-with-most-water.py
class Solution:
    def maxArea(self, height: List[int]) -> int:
        """
        [1,8,6,2,5,4,8,3,7]

        keep track of the max area with a var
        to calc area, multiply (2nd index - 1st index) * lowest height between the 2 lines

        two pointers on either end to define bounds
        whichever line is taller, move the other line up one
        """
        i, j = 0, len(height) - 1
        max_area = float('-inf')
        while i != j:
            max_area = max(max_area, (j - i) * min(height[i], height[j]))
            if height[i] < height[j]:
                i += 1
            else:
                j -= 1
            
        return max_area
        
            

File: 11-container-with-most-water/11-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {
        // return bruteForce(height);
        return optimized(height);
    }
    
    public int optimized(int[] height) {
        /*
            
            keep the larger pointer, move the smaller.
            
            [1, 8, 6, 2, 5, 4, 8, 3, 7]
                L                    R
                
            [101, 1012, 7, 99, 100]
             L                  R
        */
        
        int L = 0, R = height.length - 1;
        int maxArea = 0;
        while (L < R) {
            int area = (R - L) * Math.min(height[L], height[R]);
            maxArea = Math.max(maxArea, area);
            
            if (height[L] < height[R]) {
                L += 1;
            } else {
                R -= 1;
            }
        }
        
        return maxArea;
    }
    
    public int bruteForce(int[] height) { // TLE
        int maxArea = 0;
        for (int L = 0; L < height.length - 1; ++L) {
            for (int R = L + 1; R < height.length; ++R) {
                int area = (R - L) * Math.min(height[L], height[R]);
                maxArea = Math.max(maxArea, area);
            }
        }
        
        return maxArea;
    }
}